### DISCUSSION TITLE
Requiring a declaration to end with newline

### DISCUSSION THREAD

**Discussion by ymc9** [eclipse-langium/langium#1179] [Q&A] [answered]
Hi there,

I'm tweaking my grammar and want to ensure some of the declarations end with a new line to avoid confusion, but I'm unsure what's the proper way to do this. It seems that if I make "\n" a terminal, I have to exclude it from `WS,` then it seems I need to handle "\n" in many places.

https://github.com/eclipse-langium/langium/discussions/971

Is there an easy solution to this? Basically, in the context of the tutorial example, I want to disallow:
```
person John person Jane
```

It sounds like the recommendation is to `$cstNode` to check for newline chars (I guess that means to postpone the error reporting to the validation stage?), just want to confirm if that's the way to go.

Thanks!

**Comment by spoenemann** [comment, marked as answer]
Right, the simplest and most robust solution is probably to handle this in validation. That also makes it easier to provide a helpful error message and a quick-fix code action.

**Reply by ymc9** [reply]
Got it. Thank you for the confirmation!

**Reply by spoenemann** [reply]
This is a more general pattern: it's often a good idea to make the parser less strict, accepting some invalid inputs, to keep more control of the feedback sent to the user in the editor. When the parser is strict and the input does not conform to the syntax, you're relying fully on the parser's error recovery capabilities, which don't always behave as expected.
